language/move-vm/types/src/natives/function.rs

```diff
@@ -51,6 +53,65 @@ pub trait NativeContext {
     fn type_to_type_layout(&self, ty: &Type) -> PartialVMResult<Option<MoveTypeLayout>>;
 }
 
+pub trait NativeFunction: Copy {
+    fn run<C>(
+        &self,
+        context: &mut C,
+        ty_args: Vec<Type>,
+        args: VecDeque<Value>,
+    ) -> PartialVMResult<NativeResult>
+    where
+        C: NativeContext;
+}
+
+#[macro_export]
+macro_rules! make_enum_for_natives {
+    ($ec: ident = $($tc: ident)|*) => {
+        #[derive(Copy, Clone)]
+        pub enum $ec {
+            $($tc($tc)),*
+        }
+
+        $(impl From<$tc> for $ec {
+            fn from(x: $tc) -> $ec {
+                $ec::$tc(x)
+            }
+        })*
+
+        impl NativeFunction for $ec {
+            fn run<C>(
+                &self,
+                context: &mut C,
+                ty_args: Vec<Type>,
+                args: VecDeque<Value>,
+            ) -> PartialVMResult<NativeResult>
+            where
+                C: NativeContext,
+            {
+                match self {
+                    $($ec::$tc(f) => f.run(context, ty_args, args)),*
```

Comment:
I don't really get why this is called static dispatch. I looks like it is even more dynamic than the other approach. You have build an enum for all of the native functions, then instead of just invoking the function pointer, you have to case on which native function you are calling and get the correct `run` in the interpreter. This is is very similar to the current approach for sure, but there is nothing static about it.

Reply:
Great question! I'd say that this static vs dynamic debate is not really (just) about native function dispatching, but more about setting the tone for the VM. In other words, the goal is figure out whether the code of the VM itself should be statically or dynamically dispatched.
- This PR assumes that we'll continue using generics in the VM whenever applicable, buying into the idea of zero cost abstractions. 
- The other PR takes a more restrictive approach and tries to minimize the use of generics in the VM by taking advantage of trait objects (which are dynamically dispatched). 

> Each enum is known by the given Function... but there is no way to actually get the run function without going through a match
> Unless rust optimizes this match with a vtable-like jump setup (we did do this in Skip so it isn't out of the question), this will be slower than the function pointer solution
My assumption in this PR is that this will be a (at least somewhat) zero-cost abstraction and I'd expect either the compiler to optimize it out in the exact way you described, or the CPU to be smart enough to "fuse" the two-step jump. I'm speaking without absolute certainty since modern compilers are super complex... In fact, it can sometimes be too smart to prevent me from testing this out in an easy way -- the following two playground examples result in exactly the same ASM code, and it's not just the dispatching being optimized out -- the loop is unrolled and the whole computation is done at compile time! All the code is doing is pushing arguments (immediate operands) to the stack and call `print!`.  
- https://play.rust-lang.org/?version=stable&mode=release&edition=2018&gist=283e22b070b3a3d8e8234dbc882378bf
- https://play.rust-lang.org/?version=stable&mode=release&edition=2018&gist=24be73d3887979674bc2fc993272d6bb
Again, this may not be true for something more complex like the VM, but if there's any significant impact on performance, we should be able to detect it through benchmarks.

Reply:
>My assumption in this PR is that this will be a (at least somewhat) zero-cost abstraction and I'd expect either the compiler to optimize it out in the exact way you described, or the CPU to be smart enough to "fuse" the two-step jump

At best though, you are down to more or less having a function pointer here. 
So now we have this big abstraction moving around with generics and then hope that Rust optimizes it. There is no way it will completely optimize away the enum itself as it does on the playground as these things are being done dynamically. There will be no static dispatch here when done for the VM.

At this point I really don't care about benchmarks because it will likely all be noise; natives just aren't used that extensively and even then the best case here is the compiler optimizes this solution into the function pointer one. 
We should just go with the function pointer solution because it of how much simpler it is.

Reply:
@tnowacki I think there's some misunderstanding here (mostly due to my choice of words being ambiguous) By "dynamic vs static" I'm not really referring to the way native functions are dispatched, but rather how code/components (logger, storage etc.) in the VM should be dispatched. In other words, I want to figure out whether in general, we should prefer generics, or trait objects in the VM, and depending on our answer to that question, one of the native function implementations will likely stand out due to limitations of the language. IMO our choice here will set the tone for the VM for the foreseeable future, so I want to study both styles a little bit more carefully. 

Right now I'm done with some of the simple benchmarks (with the two current implementations unmodified) and I see no practical difference in terms of performance. At this point I think we should move forward with the function pointer based implementation for extensible native functions, in favor of its simplicity. With that being said, I will continue conducting some additional benchmarks, tweaking the code a bit here and there (e.g. switching storage to trait objects) and see if there are performance impacts to help us understand more about the performance characteristics of the VM and make better decisions in the future.
```
group           base                                   dynamic-1                              dynamic-2                              dynamic-3                              dynamic-4                              dynamic-5                              dynamic-6                              static-1                               static-2                               static-3                               static-4                               static-5                               static-6
-----           ----                                   ---------                              ---------                              ---------                              ---------                              ---------                              ---------                              --------                               --------                               --------                               --------                               --------                               --------
arith           1.01     21.7±0.89ms        ? ?/sec    1.01     21.8±0.67ms        ? ?/sec    1.03     22.2±1.00ms        ? ?/sec    1.00     21.6±0.99ms        ? ?/sec    1.04     22.5±0.89ms        ? ?/sec    1.01     21.9±0.92ms        ? ?/sec    1.02     22.1±1.06ms        ? ?/sec    1.04     22.5±0.61ms        ? ?/sec    1.01     21.7±0.96ms        ? ?/sec    1.00     21.6±1.34ms        ? ?/sec    1.01     21.8±0.53ms        ? ?/sec    1.03     22.4±0.83ms        ? ?/sec    1.01     21.9±0.99ms        ? ?/sec
call            1.05     19.6±1.02ms        ? ?/sec    1.03     19.2±0.64ms        ? ?/sec    1.02     19.1±0.84ms        ? ?/sec    1.03     19.3±0.89ms        ? ?/sec    1.08     20.0±0.96ms        ? ?/sec    1.03     19.2±0.75ms        ? ?/sec    1.05     19.5±0.81ms        ? ?/sec    1.00     18.7±0.90ms        ? ?/sec    1.00     18.6±0.99ms        ? ?/sec    1.00     18.6±0.80ms        ? ?/sec    1.06     19.7±1.26ms        ? ?/sec    1.03     19.2±0.89ms        ? ?/sec    1.01     18.7±1.01ms        ? ?/sec
natives         1.05     25.7±0.54ms        ? ?/sec    1.00     24.5±0.89ms        ? ?/sec    1.00     24.4±1.00ms        ? ?/sec    1.03     25.1±2.29ms        ? ?/sec    1.03     25.0±1.28ms        ? ?/sec    1.03     25.2±0.67ms        ? ?/sec    1.01     24.7±0.89ms        ? ?/sec    1.01     24.8±0.93ms        ? ?/sec    1.01     24.7±1.30ms        ? ?/sec    1.01     24.7±0.71ms        ? ?/sec    1.02     24.8±0.81ms        ? ?/sec    1.04     25.4±1.39ms        ? ?/sec    1.04     25.3±1.32ms        ? ?/sec
peer_to_peer    1.08   276.3±17.56ms        ? ?/sec    1.05    269.8±7.43ms        ? ?/sec    1.03    263.4±7.45ms        ? ?/sec    1.02    262.0±7.49ms        ? ?/sec    1.04    265.8±6.93ms        ? ?/sec    1.03    263.8±7.90ms        ? ?/sec    1.03    263.6±8.67ms        ? ?/sec    1.05    269.4±7.40ms        ? ?/sec    1.00    256.0±7.21ms        ? ?/sec    1.03    263.2±8.97ms        ? ?/sec    1.04    267.1±8.13ms        ? ?/sec    1.02   261.1±10.94ms        ? ?/sec    1.03    262.7±7.48ms        ? ?/sec
```